### PR TITLE
Add ConditionEnvironment= and ConditionIsEncrypted=

### PR DESCRIPTION
**Pull request by poettering**
Prompted by #15180

### CODE REVIEW

File: man/systemd.unit.xml [hunk 1]
```diff
@@ -1346,6 +1362,18 @@
...
 
+        <varlistentry>
+          <term><varname>ConditionPathIsEncrypted=</varname></term>
+
+          <listitem><para><varname>ConditionPathIsEncrypted=</varname> is similar to
+          <varname>ConditionPathIsReadWrite=</varname> but verifies that the underlying file system's backing
+          block device is encrypted using dm-crypt/LUKS. Note that this check does not cover ext4
+          per-directory encryption, and only detects block level encryption. Moreover, while dm-crypt/LUKS
+          encryption on top of loopback block devices is detected, whether the loopback file below a loopback
+          block device resides on an encrypted block device is not.</para>
```
Comment (keszybz) [inline man/systemd.unit.xml]: I read that last sentence like 10 times and I still have no idea what it means. Turtles all the way down ;)

File: man/systemd.unit.xml [hunk 2]
```diff
@@ -1226,6 +1226,22 @@
...
+        <varlistentry>
+          <term><varname>ConditionEnvironment=</varname></term>
+
+          <listitem><para><varname>ConditionEnvironment=</varname> may be used to check whether a specific
+          environment variable is set (or if prefixed with the exclamation mark — unset) in the service
+          manager's environment block. The argument must either be a single word, or an assignment (i.e. two
+          words, separated by <literal>=</literal>). In the former case the environment is searched as left
+          hand side of an environment variable assignment. In the latter case, the exact assignment is looked
+          for with right and left hand side matching. Note that the environment block of the service manager
+          itself is checked, i.e. not any variables defined with <varname>Environment=</varname> or
```
Comment (keszybz) [inline man/systemd.unit.xml]: Maybe: "The argument may be a single word, to check if the variable with this name is defined in the environment block, or an assignment (<literal>name=value</literal>), to check if the variable with this exact value is defined."

File: src/core/unit.c [hunk 3]
```diff
@@ -1707,24 +1707,50 @@ static int log_unit_internal(void *userdata, int level, int error, const char *f
...
+        _cleanup_strv_free_ char **env = NULL;
+        int r;
+
         assert(u);
 
         dual_timestamp_get(&u->assert_timestamp);
-        u->assert_result = condition_test_list(u->asserts, assert_type_to_string, log_unit_internal, u);
 
-        unit_add_to_dbus_queue(u);
+        r = manager_get_effective_environment(u->manager, &env);
```
Comment (keszybz) [inline src/core/unit.c:1739]: Hmm, this will happen quite a lot. Maybe we should cache the effective environment?

File: src/shared/condition.c [hunk 4]
```diff
@@ -625,55 +625,84 @@ static int condition_test_first_boot(Condition *c) {
...
+        STRV_FOREACH(i, environ) {
+                bool found;
+
+                if (equal)
+                        found = streq(c->parameter, *i);
+                else {
+                        const char *f;
+
+                        f = startswith(*i, c->parameter);
+                        found = f && IN_SET(*f, 0, '=');
```
Comment (keszybz) [inline src/shared/condition.c:647]: I was wondering whether this part is necessary, i.e. whether a variable in the from not matching k=v is possible. But environ(7) says
> By convention the strings in environ have the form "name=value".

which means that anything is possible. Wow.

File: src/core/unit.c (hunk 3 above)
Comment (poettering) [reply src/core/unit.c:1739]: we certainly can, but it's definitely material for another PR. and we might want to do some profiling first.

File: man/systemd.unit.xml (hunk 1 above)
Comment (poettering) [reply man/systemd.unit.xml]: ok, i rewrote that sentence, and made it two.

File: man/systemd.unit.xml [hunk 5]
```diff
@@ -1226,6 +1226,24 @@
...
+          manager's environment block.
+
+          The argument may be a single word, to check if the variable with this name is defined in the
+          environment block, or an assignment
+          (<literal><replaceable>name</replaceable>=<replaceable>value</replaceable></literal>), to check if
+          the variable with this exact value is defined. Note that the environment block of the service
+          manager itself is checked, i.e. not any variables defined with <varname>Environment=</varname> or
+          <varname>EnvironmentFile=</varname>, as described above. This is particularly useful when the
+          service manager runs inside a containerized environment or as per-user service manager, in order to
+          check for variables passed in by the enclosing container manager or PAM.</para>
```
Comment (keszybz) [inline man/systemd.unit.xml:1243]: I wonder if we this could be of use for #15226...

**Pull request merged**